File: leetcode/2130 Maximum Twin Sum of a Linked List/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func pairSum(head *ListNode) int { // 快慢指针 + 反转链表！！！
    slow, fast := head, head.Next
    for fast.Next != nil && fast.Next.Next != nil { // fast到达链表末尾，slow刚好指向前一半部分链表的末尾节点！！！
        slow = slow.Next
        fast = fast.Next.Next
    }

    head2 := slow.Next
    slow.Next = nil

    var previous, current *ListNode = nil, head2 // 见206题反转链表！！！
    for (current != nil) {
        var nextNode *ListNode = current.Next
        current.Next = previous
        previous = current
        current = nextNode
    }

    result := math.MinInt32
    for head != nil {
        result = max(result, head.Val + previous.Val)
        head, previous = head.Next, previous.Next
    }
    return result
}

func max(a int, b int) int {
    if a >= b {
        return a
    } else {
        return b
    }
}

File: leetcode/2130 Maximum Twin Sum of a Linked List/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) { // 快慢指针 + 反转链表！！！
        ListNode slow = head;
        ListNode fast = head.next;
        while (fast.next != null && fast.next.next != null) { // fast到达链表末尾，slow刚好指向前一半部分链表的末尾节点！！！
            slow = slow.next;
            fast = fast.next.next;
        }
        
        ListNode head2 = slow.next;
        slow.next = null;

        ListNode previous = null; // 见206题反转链表！！！
        ListNode current = head2;
        while (current != null){
            ListNode nextNode = current.next;
            current.next = previous;
            previous = current;
            current = nextNode;
        }

        int result = Integer.MIN_VALUE;
        while (head != null) {
            result = Math.max(result, head.val + previous.val);
            head = head.next;
            previous = previous.next;
        }
        return result;
    }
}